internal/wal/wal_aof.go

```diff
@@ -65,31 +170,212 @@ func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
 	data, err := proto.Marshal(entry)
 	if err != nil {
 		slog.Warn("failed to serialize command", slog.Any("error", err.Error()))
+		return
 	}
 
-	writeBuf.Reset()
-	writeBuf.Grow(4 + len(data))
-	if binary.Write(&writeBuf, binary.BigEndian, uint32(len(data))) != nil {
-		slog.Warn("failed to write entry length to WAL", slog.Any("error", err.Error()))
+	buf := bufferPool.Get().([]byte)
+	buf = buf[:0]
+	defer bufferPool.Put(buf)
+
+	length := uint32(len(data))
+	totalSize := 4 + len(data)
+
+	if cap(buf) < totalSize {
+		buf = make([]byte, totalSize)
+	} else {
+		buf = buf[:totalSize]
+	}
+
+	binary.BigEndian.PutUint32(buf[0:4], length)
+	copy(buf[4:], data)
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	if w.rotationMode == "size" {
+		if err := w.rotateLogIfNeeded(); err != nil {
+			slog.Warn("failed to rotate WAL log", slog.Any("error", err.Error()))
+			return
+		}
 	}
-	writeBuf.Write(data)
 
-	if _, err := w.file.Write(writeBuf.Bytes()); err != nil {
+	if _, err := w.writer.Write(buf); err != nil {
 		slog.Warn("failed to write serialized command to WAL", slog.Any("error", err.Error()))
+		return
 	}
 
-	if err := w.file.Sync(); err != nil {
-		slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+	if w.writeMode == "fsync" {
+		if w.isBuffered && w.bufWriter != nil {
+			if err := w.bufWriter.Flush(); err != nil {
+				slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+				return
+			}
+		}
+		if err := w.file.Sync(); err != nil {
+			slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+			return
+		}
+	} else {
+		w.syncMutex.Lock()
+		defer w.syncMutex.Unlock()
+		if time.Since(w.lastSyncTime) >= w.syncInterval {
+			if w.isBuffered && w.bufWriter != nil {
+				if err := w.bufWriter.Flush(); err != nil {
+					slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+					return
+				}
+			}
+			if err := w.file.Sync(); err != nil {
+				slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+				return
+			}
+			w.lastSyncTime = time.Now()
+		}
 	}
 
 	slog.Debug("logged command in WAL", slog.Any("command", c.Repr()))
 }
 
-func (w *WALAOF) Close() error {
-	if w.file == nil {
+func (w *WALAOF) rotateLogIfNeeded() error {
+	fileInfo, err := w.file.Stat()
+	if err != nil {
+		return err
+	}
+
+	var bufferedSize int64 = 0
+	if w.isBuffered && w.bufWriter != nil {
+		bufferedSize = int64(w.bufWriter.Buffered())
+	}
+
+	if fileInfo.Size()+bufferedSize >= w.maxFileSize {
+		return w.rotateLog()
+	}
+	return nil
+}
+
+func (w *WALAOF) rotateLog() error {
```

Comment:
why do you need this function? this was not part of the abstract class anyway. Close and Init should do the trick. We should not require any additional function part of the WALAOF